frontend/public/components/build-pipeline.tsx

```diff
@@ -0,0 +1,174 @@
+  </div> : null;
+};
+
+const StagesNotStarted: React.SFC<StagesNotStartedProps> = ({stages}) => {
+  return !stages.length ? <div className="build-pipeline__stage build-pipeline__stage--none">No stages have started.
+  </div> : null;
+};
+
+
+const BuildSummaryTimestamp: React.SFC<BuildSummaryTimestampProps> = ({timestamp}) => <span className="build-pipeline__timestamp text-muted">
+  { fromNow(timestamp) }
+</span>;
+
+const BuildPipelineSummary: React.SFC<BuildPipelineSummaryProps> = ({obj}) => {
+  const buildNumber = getBuildNumber(obj);
+  return <div className="build-pipeline__summary">
+    <div className="build-pipeline__phase">
+      <BuildSummaryStatusIcon status={obj.status.phase} /> Build {buildNumber}
+    </div>
+    <BuildSummaryTimestamp timestamp={obj.metadata.creationTimestamp} />
+    <BuildLogLink obj={obj} />
+  </div>;
+};
+
+const BuildAnimation:React.SFC<BuildAnimationProps> = ({status}) => <div className={`build-pipeline__status-bar build-pipeline__status-bar--${_.kebabCase(status)}`}>
+  <div className="build-pipeline__animation-line"></div>
+  <div className="build-pipeline__animation-circle">
+    <div className="build-pipeline__circle-clip1"></div>
+    <div className="build-pipeline__circle-clip2"></div>
+    <div className="build-pipeline__circle-inner">
+      <div className="build-pipeline__circle-inner-fill"></div>
+    </div>
+  </div>
+</div>;
+
+const JenkinsInputUrl: React.SFC<JenkinsInputUrlProps> = ({obj, stage}) => {
+  const pending = stage && (stage.status === 'PAUSED_PENDING_INPUT');
+  const buildUrl = getJenkinsBuildURL(obj);
+  if(pending) {
+    return <div className="build-pipeline__stage-actions text-muted">
+      <a href={buildUrl} target="_blank" rel="noopener noreferrer">Input Required</a>
+    </div>;
+  }
+  return null;
+};
+
+const BuildStageTimestamp:React.SFC<BuildStageTimestampProps> = ({timestamp}) => <div className="build-pipeline__stage-time text-muted">
+  { fromNow(timestamp) }
+</div>;
+
+const BuildStageName: React.SFC<BuildStageNameProps> = ({name}) => {
+  return <div title={name} className="build-pipeline__stage-name">
+    {name}
+  </div>;
+};
+
+const BuildStage: React.SFC<BuildStageProps> = ({obj, stage}) => {
+  return <div className="build-pipeline__stage">
+    <div className="build-pipeline__stage-column">
+      <BuildStageName name={stage.name} />
+      <BuildAnimation status={stage.status} />
+      <JenkinsInputUrl obj={obj} stage={stage} />
+      <BuildStageTimestamp timestamp={stage.startTimeMillis} />
+    </div>
+  </div>;
+};
+
+export const BuildPipeline: React.SFC<BuildPipelineProps> = ({obj}) => {
+  const jenkinsStatus: any = getJenkinsStatus(obj);
+  const stages = getStages(jenkinsStatus);
+  return <div className="build-pipeline">
+    <BuildPipelineSummary obj={obj} />
+    <div className="build-pipeline__container">
+      <div className="build-pipeline__stages">
+        <StagesNotStarted stages={stages} />
```

Comment:
This is a little confusing because it looks like `StagesNotStarted` is always visible, but then you loop over stages below. I think it's much clearer if you didn't use the component:

```jsx
{_.isEmpty(stages)
  ? <div className="build-pipeline__stage build-pipeline__stage--none">No stages have started.</div>
  : stages.map(stage => <BuildStage obj={obj} stage={stage} key={stage.id} /> )}
```

Reply:
I can put the `_.isEmpty()` check up here, that makes sense.